### DISCUSSION TITLE
希望能从社区挑选出一些最佳实践，方便工程化上手Tauri

### DISCUSSION THREAD

**Discussion by jackhb1999** [tauri-apps/tauri#11903] [General]
由于我的rust学习经验非常浅薄，但是结合可视化界面学习是一种能有成就感的方式，这是我愿意选择Tauri做工程的原因。
但是这个过程中遇到了一些问题，由于很多内容不了解总会走到弯路。
希望有些优秀实践作为我学习过程的指路明灯。

另外留下我学习过程中的一个问题：
我在做 Tauri 结合tokio、sea-orm遇到的问题，启动应用，第一次连接数据库时会连接报错，但是会自动重连，我的数据库工作依然继续，但是这个连接问题还是让我困惑。
![image](https://github.com/user-attachments/assets/db42e7ca-5877-49a6-8d74-e98d6d1273aa)
`我不确定这是一个bug，还是我实践中的使用有问题。
```
#[cfg_attr(mobile, tauri::mobile_entry_point)]
pub fn run() {
    tokio::runtime::Builder::new_current_thread()
        .enable_all()
        .build()
        .unwrap()
        .block_on(async move {
            let db = handle::ConnectionHandle::establish_connection().await.unwrap();
            tauri::Builder::default()
                .plugin(tauri_plugin_shell::init())
                .manage(db)
                .invoke_handler(tauri::generate_handler![greet,login,getGoodsList,getUserList,getRoleList])
                .run(tauri::generate_context!())
                .expect("error while running tauri application");
        });
}

pub async fn establish_connection() -> anyhow::Result<DatabaseConnection> {
    let db_url = "postgres://postgres:fackpg@localhost/test";
    let conn = Database::connect(db_url).await
        .expect("connection established: 连接出错");
    println!("已创建连接");
    Ok(conn)
}

#[tauri::command]
async fn login(name: &str, pass: &str, db: State<'_, DatabaseConnection>) -> Result<String,String> {
    println!("Logging in...{}", pass);
    println!("Logging in...{:?}", db);
    handle::LogonHandle::handle(name, pass, db).await.expect("登录失败！");
    // 登录
   Ok( format!("Hello, {}! You've been greeted from Rust!{}", name, pass))
}

pub async fn handle(name: &str, pass: &str, db: State<'_, DatabaseConnection>) -> anyhow::Result<()> {
            let conn = db.inner();
            println!("获取连接");

// 其他代码

}

```
我不知道是否应该留下仓库地址，但是我真的很需要引导。

**Comment by FabianLars** [comment]
First of all, i'd recommend to rewrite your run() function a bit so that tauri isn't spawned in a single threaded tokio runtime. Also, tauri already provides access to its own tokio runtime so you can just use that:
```rs
#[cfg_attr(mobile, tauri::mobile_entry_point)]
pub fn run() {
    let db = tauri::async_runtime::block_on(handle::ConnectionHandle::establish_connection()).unwrap();

    tauri::Builder::default()
            .plugin(tauri_plugin_shell::init())
            .manage(db)
            .invoke_handler(tauri::generate_handler![greet,login,getGoodsList,getUserList,getRoleList])
            .run(tauri::generate_context!())
            .expect("error while running tauri application");
}
```

This will probably not solve the problem but it's still worth a try. If indeed does not help i would look into https://docs.rs/sea-orm/latest/sea_orm/struct.ConnectOptions.html#method.acquire_timeout and https://docs.rs/sea-orm/latest/sea_orm/struct.ConnectOptions.html#method.connect_timeout in case it really is just a timeout.
If this also doesn't help, then i would recommend to reach out to the sea-orm or sqlx communities. They may be able to help you better than us.

**Reply by jackhb1999** [reply]
> First of all, i'd recommend to rewrite your run() function a bit so that tauri isn't spawned in a single threaded tokio runtime. Also, tauri already provides access to its own tokio runtime so you can just use that:
> 
> ```rust
> #[cfg_attr(mobile, tauri::mobile_entry_point)]
> pub fn run() {
>     let db = tauri::async_runtime::block_on(handle::ConnectionHandle::establish_connection()).unwrap();
> 
>     tauri::Builder::default()
>             .plugin(tauri_plugin_shell::init())
>             .manage(db)
>             .invoke_handler(tauri::generate_handler![greet,login,getGoodsList,getUserList,getRoleList])
>             .run(tauri::generate_context!())
>             .expect("error while running tauri application");
> }
> ```
> 
> This will probably not solve the problem but it's still worth a try. If indeed does not help i would look into https://docs.rs/sea-orm/latest/sea_orm/struct.ConnectOptions.html#method.acquire_timeout and https://docs.rs/sea-orm/latest/sea_orm/struct.ConnectOptions.html#method.connect_timeout in case it really is just a timeout. If this also doesn't help, then i would recommend to reach out to the sea-orm or sqlx communities. They may be able to help you better than us.

非常感谢您的回答，而且这真的解决了我现在遇到的问题。
关于连接超时，这个确实是原有代码100%可以复现的问题。
当我不使用 tauri 的 manage 管理我的数据库连接，而是在对应的 command 方法中使用 tokio 的运行时进行阻塞创建及使用，完全没有问题。
我尝试用 tauri 的 manage 管理我的数据库连接，并使用 tokio 运行时，这好像和 tauri 自己的运行时打架了，于是会有超时重连的情况。
当我使用你建议的 tauri::async_runtime 则完全不会出现问题，也许我应该将直接使用 tokio 的所有地方替换为 tauri::async_runtime。
